### PR TITLE
EAHW-1707: Night Hours - Count

### PR DESCRIPTION
**Pull request by joannafawl**
- Created NightHoursAccrualsModule
- Moved logic for `calculateShiftContribution` into individual accruals modules
- Extended integration tests to include night hours

Note: the refactoring work will be done in a subsequent PR

### CODE REVIEW

File: src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculatorIntegrationTest.java [hunk 1]
```diff
@@ -43,27 +45,57 @@ void calculate_endToEndInGmtToBst_contributionsAndCumulativeTotalsAsExpected() {
...
         .usingComparator(BigDecimal::compareTo).isEqualTo(BigDecimal.valueOf(240));
     assertThat(accruals.get(2).getCumulativeTotal())
         .usingComparator(BigDecimal::compareTo).isEqualTo(BigDecimal.valueOf(7440));
 
+    assertThat(accruals.get(3).getAccrualType()).isEqualTo(ANNUAL_TARGET_HOURS);
     assertThat(accruals.get(3).getContributions().getTotal())
         .usingComparator(BigDecimal::compareTo).isEqualTo(BigDecimal.valueOf(720));
     assertThat(accruals.get(3).getCumulativeTotal())
         .usingComparator(BigDecimal::compareTo).isEqualTo(BigDecimal.valueOf(8160));
+
```
Comment (stephenkaye-ho) [inline src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculatorIntegrationTest.java:74]: Could we make this a method i.e.
```
assertTypeAndTotals(accrual, expectedAccrualType, expectedContributionTotal, expectedCumulativeTotal)
```

File: src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        .map(range -> new BigDecimal(
+            Duration.between(range.lowerEndpoint(), range.upperEndpoint()).toMinutes()))
+        .reduce(BigDecimal.ZERO, BigDecimal::add);
+  }
+
+  private RangeSet<ZonedDateTime> getIntersectionWithNightHours(ZonedDateTime startTime,
+                                                                ZonedDateTime endTime) {
+
+    ZonedDateTime midnightCurrentDay = atTime(startTime, 0);
+    ZonedDateTime endOfPreviousDayNightHours = atTime(startTime, 6);
```
Comment (stephenkaye-ho) [inline src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:37]: Would it be useful to have these values as constants? Thinking just for prominence if nothing else.

File: src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.enums.AccrualType;
+
+public class NightHoursAccrualModule extends AccrualModule {
+
+  public NightHoursAccrualModule() {
+    super(AccrualType.NIGHT_HOURS);
+  }
+
+  @Override
+  public BigDecimal calculateShiftContribution(ZonedDateTime startTime, ZonedDateTime endTime) {
```
Comment (stephenkaye-ho) [inline src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:20]: I thought we checked for having javadocs on public methods. I was looking to more easily understand what the private method was doing which wouldn't be covered in this case but then realised that our javadocs have disappeared.

File: src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModuleTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        Arguments.of(ZonedDateTime.parse("2023-03-18T00:00:00+00:00"),
+            ZonedDateTime.parse("2023-03-18T06:00:00+00:00"),
+            BigDecimal.valueOf(360)),
+        Arguments.of(ZonedDateTime.parse("2023-03-18T04:00:00+00:00"),
+            ZonedDateTime.parse("2023-03-18T10:00:00+00:00"),
+            BigDecimal.valueOf(120)),
+        Arguments.of(ZonedDateTime.parse("2023-03-18T14:00:00+00:00"),
+            ZonedDateTime.parse("2023-03-18T23:30:00+00:00"),
+            BigDecimal.valueOf(30)),
+        Arguments.of(ZonedDateTime.parse("2023-03-18T23:30:00+00:00"),
```
Comment (stephenkaye-ho) [inline src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModuleTest.java:31]: I'm assuming this wasn't intended to quite be a boundary test but it does highlight that we don't have one completing at midnight.
```
        Arguments.of(ZonedDateTime.parse("2023-03-18T23:30:00+00:00"),
            ZonedDateTime.parse("2023-03-19T00:00:00+00:00"),
            BigDecimal.valueOf(30)),
```

**Review by stephenkaye-ho — CHANGES_REQUESTED** [review]

**Review by IanFisherHO — APPROVED** [review]
Just an observation. As we add more accrual types we might want to start generating the test data in loops or splitting the data down into smaller files per accrual type.

File: src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculatorIntegrationTest.java (hunk 1 above)
Comment (joannafawl) [reply src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculatorIntegrationTest.java:56]: Updated tests to use `assertTypeAndTotals` method

File: src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java (hunk 3 above)
Comment (joannafawl) [reply src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:31]: Added a doc comment to the method

File: src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModuleTest.java (hunk 4 above)
Comment (joannafawl) [reply src/test/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModuleTest.java:31]: Added another test scenario for the midnight boundary

File: src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java [hunk 5]
```diff
@@ -12,10 +12,21 @@
 
 public class NightHoursAccrualModule extends AccrualModule {
 
+  private static final int MIDNIGHT = 0;
```
Comment (stephenkaye-ho) [inline src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:15]: My fault on this, I was meaning something like
NIGHT_HOURS_START=23
NIGHT_HOURS_END=6
Then if they ever changed to 10pm-5am we'd only need to change the constant value not it's name

File: src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java (hunk 3 above)
Comment (stephenkaye-ho) [reply src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:31]: More to my point was that this should probably not have been copied and pasted into this repo
https://github.com/UKHomeOffice/callisto-balance-calculator/blob/main/checkstyle_suppressions.xml

Comment (stephenkaye-ho) [reply src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/module/NightHoursAccrualModule.java:31]: Discussed with @joannafawl  and agreed it's a topic wider than just this PR so will resolve this conversation without further action

**Review by JWeston-HO — APPROVED** [review]
